### PR TITLE
feat(labs): New compound toast component

### PR DESCRIPTION
**Pull request by ckaptan**
## Summary

Resolves: #1199  

Adds toast compound component. New component adheres to the look and feel of the existing toast component. README and stories provide further examples about component's structure.

category

### Release Note
We've converted Toast to a compound component. As part of this change, we've:
- Added accessibility attributes based on the `mode` of the Toast.
- Cleaned up the action link to be a `HyperLink`
- Used our `Popup` subcomponents to build out Toast components
- Created a `useToastModel` that provides access to `mode` and `id`

---

## Checklist

- [x] tests are changed or added
- [x] code has been documented
- [x] design approved final implementation
- [x] a11y approved final implementation
- [x] code adheres to the [API & Pattern guidelines](https://workday.github.io/canvas-kit/?path=/story/welcome-dev-docs-api-pattern-guidelines--page)

### CODE REVIEW

File: modules/labs-react/package.json
Comment (willklein) [inline modules/labs-react/package.json:51]: This is fine but let's bump it up to be consistent. It resolves to `0.2.8 `anyway.
```suggestion
    "@workday/design-assets-types": "^0.2.8",
```

File: modules/labs-react/toast/lib/ToastMessage.tsx
Comment (RayRedGoose) [inline modules/labs-react/toast/lib/ToastMessage.tsx:18]:
```suggestion
  Component: (elemProps: ToastMessageProps, ref, Element) => {
    return <Message as={Element} ref={ref}  {...elemProps} />;
  },
```

File: modules/labs-react/toast/lib/Toast.tsx
Comment (RayRedGoose) [inline modules/labs-react/toast/lib/Toast.tsx:39]: how you feel to replace `mode === 'interactive'` and `mode === 'assertive'` statements by variables? Or adding a function to generate specific attributes for each mode? Something like this: 
```
const getAccessibleAttributes = (mode) => {
  switch (mode) {
    case "interactive":
      return {
        "aria-describedby": randomDescribedbyID,
        "aria-label": 'notification',
        role: 'dialog',
      }
    case "assertive":
      return {
        role: 'alert',
        "aria-live": 'assertive'
      }
    case "polite":
      return {
        role: 'status',
        "aria-live": 'polite',
        'aria-atomic': true
      }
  }
}
```

Here we have 4 conditional statement for this, maybe it's better to turn it into just one.

Comment (mannycarrera4) [reply modules/labs-react/toast/lib/Toast.tsx:39]: i like this!

File: modules/labs-react/toast/stories/stories_VisualTesting.tsx
Comment (RayRedGoose) [inline modules/labs-react/toast/stories/stories_VisualTesting.tsx:25]: It seems like this modes should be changed too.

File: modules/labs-react/toast/lib/ToastIcon.tsx
Comment (mannycarrera4) [reply modules/labs-react/toast/lib/ToastIcon.tsx:10]: good catch, I'll ad an `Omit`

File: modules/labs-react/toast/lib/ToastCloseIcon.tsx
Comment (mannycarrera4) [reply modules/labs-react/toast/lib/ToastCloseIcon.tsx:14]: nope, will clean this up!

File: modules/labs-react/toast/README.md
Comment (alanbsmith) [inline modules/labs-react/toast/README.md:8]:
```suggestion
Toasts contain low-emphasis updates or messages about the status of an application’s process.
```
Nit: I'm going to [borrow language](https://canvas.workdaydesign.com/components/popups/toast/) from our design site instead.

Comment (jamesfan) [reply modules/labs-react/toast/README.md]: I ended up wiping out most of the README.md as is convention after migrating a component's docs to MDX. I did include this language in the MDX, however.

File: modules/labs-react/toast/lib/Toast.tsx
Comment (jamesfan) [inline modules/labs-react/toast/lib/Toast.tsx:61]: The vertical alignment on the ToastIcon seems off compared with the ToastMessage:

icon-vertical-alignment

It seems like the message needs to be a little further down to be vertically centered with the icon.

File: modules/labs-react/toast/stories/examples/RTL.tsx
Comment (jamesfan) [inline modules/labs-react/toast/stories/examples/RTL.tsx:15]: Some of our other examples nest the message within `Toast.Message`. Here (and in a few other examples as well), the message is contained directly within `Toast.Body` -- should we be using `<Toast.Body><Toast.Message>...</Toast.Message></Toast.Body>` here as well?

**Review by jamesfan — COMMENTED** [review]
I made a first pass at cleaning up the docs to match our [MDX template](https://github.com/Workday/canvas-kit/discussions/1131). Had a few questions/suggestions which I've included below.

File: modules/labs-react/toast/spec/ToastClose.spec.tsx
Comment (alanbsmith) [inline modules/labs-react/toast/spec/ToastClose.spec.tsx:6]:
```suggestion
describe('Toast.CloseIcon', () => {
```

File: modules/labs-react/toast/stories/examples/Basic.tsx
Comment (alanbsmith) [inline modules/labs-react/toast/stories/examples/Basic.tsx:9]: Comment (not an RC): It would be nice if this prop was a token-aware, so I could do this:

```tsx
<Toast.Icon icon={checkIcon} color="greenApple400" />
```

But we'd need to make some adjustments to `SystemIcon` for that to work.

- Adjust the color type in `SystemIconStyles`
  - `string` becomes `CanvasColor | (string & {});`
- Remove `color` from the Omit props
- Add a function to process the colors

```ts
import {iconColors, colors, CanvasColor} from '@workday/canvas-kit-react/tokens';

function getColor(value: CanvasColor | string) {
  if (value in colors) {
    return colors[value as keyof typeof colors];
  }
  return value;
}
```

File: modules/labs-react/toast/spec/ToastClose.spec.tsx
Comment (alanbsmith) [inline modules/labs-react/toast/spec/ToastClose.spec.tsx:5]: Nit: This file should be renamed to `ToastCloseIcon.spec.tsx`

**Review by alanbsmith — CHANGES_REQUESTED** [review]
Really nice work here, @ckaptan and @mannycarrera4. 👏🏻 👏🏻 👏🏻 

I made quite a few comments and suggestions, most of which are small, nothing major. And I left some questions too. I try to preface every comment with a label so my intent is more clear.

- RC: (Request change) This should change before this can be approved
- Suggestion: This should change, and here's a snippet of how I would do it
  - Feel free to use it, or go about it your own way, but please address the issue
- Nit: Requesting a small (usually stylistic) change
  - Make the change if you want, but it's not blocking the PR
- Question: I'm looking for more context about what I'm seeing.
  - Usually not asking for a change, just clarification
- Comment: I'm just running my mouth and making observations
  - Absolutely not asking for a change to be made
  
Hopefully, that's helpful. Thanks again for working on this. This component is looking great, and I'm really excited to get it in the library! 🥳

File: modules/labs-react/toast/stories/toast.stories.mdx
Comment (mannycarrera4) [reply modules/labs-react/toast/stories/toast.stories.mdx:65]: Based on my talks with accessibility, role`alert` will read out the message in the toast and yes, both status and alert would go in a live region area! The dialog needs the describedby so that it reads the text of the toast that is associated with the toast. Here's what the accessibility tree looks like for dialog.

Screen Shot 2022-08-29 at 8 23 09 AM

File: modules/labs-react/toast/lib/ToastCloseIcon.tsx
Comment (mannycarrera4) [reply modules/labs-react/toast/lib/ToastCloseIcon.tsx:9]: https://github.com/Workday/canvas-kit/issues/1746

File: modules/labs-react/toast/stories/examples/Basic.tsx
Comment (mannycarrera4) [reply modules/labs-react/toast/stories/examples/Basic.tsx:9]: https://github.com/Workday/canvas-kit/issues/1751

File: modules/labs-react/toast/lib/ToastBody.tsx
Comment (mannycarrera4) [reply modules/labs-react/toast/lib/ToastBody.tsx:20]: ` Popup.Card` uses `Card.Body` which is a `Box` component. Those flex properties don't exist on Box 😞

Comment (RayRedGoose) [reply modules/labs-react/toast/lib/ToastBody.tsx:21]: But you can make it as `styled(Popup.Body.as(Flex))`

Comment (mannycarrera4) [reply modules/labs-react/toast/lib/ToastBody.tsx:21]: Yep, i tried that but I was getting an error and it was not loading

Comment (mannycarrera4) [reply modules/labs-react/toast/lib/ToastBody.tsx:21]: `Unexpected error while loading ./labs-react/toast/stories/toast.stories.mdx: TypeError: _canvasKitReact.Popup.Body.as is not a function`

Comment (RayRedGoose) [reply modules/labs-react/toast/lib/ToastBody.tsx:21]: Weird. I can do this on my machine 🤔

Screen Shot 2022-08-29 at 10 49 35 AM

Comment (RayRedGoose) [reply modules/labs-react/toast/lib/ToastBody.tsx:21]: Okay, it seems like `createSubcomponent` and `createContainer` don't allow us to use `.as` in styled functions the same way as `createComponent` does.

File: modules/labs-react/toast/lib/ToastCloseIcon.tsx
Comment (alanbsmith) [reply modules/labs-react/toast/lib/ToastCloseIcon.tsx:10]: Thank you!

File: modules/labs-react/toast/stories/toast.stories.mdx
Comment (jamesfan) [inline modules/labs-react/toast/stories/toast.stories.mdx:87]:
```suggestion
### Dialog

Set the `mode` prop to `dialog` to convey the presence of a follow-up action.

<ExampleCodeBlock code={ToastDialog} />

The `Toast` will be given a role of `dialog` along with an `aria-label` to indicate that it's a
notification.

> **Note**: You must use the `Toast.Message` subcomponent when the `mode` is `dialog` so that `id`
> is tied to the message for accessibility.

You can also pass in a `model` with `useToastModel` to provide a specific `id` for the
`Toast.Message`:

```tsx
export const CustomIDToast = () => {
  const model = useToastModel({
    id: '12df5',
    mode: 'dialog',
  });
  return (
    <Toast model={model}>
      <Toast.Icon icon={checkIcon} color={colors.greenApple400} />
      <Toast.Body>
        <Toast.Message>Your workbook was successfully processed.</Toast.Message>
        <Toast.Link href="#href">View More Details</Toast.Link>
      </Toast.Body>
    </Toast>
  );
};
```
```

**Review by alanbsmith — APPROVED** [review]
Looks great! Thanks, @ckaptan and @mannycarrera4!

**Review by jamesfan — APPROVED** [review]

**Pull request merged**
